### PR TITLE
Feature/363 configuration layer

### PR DESCRIPTION
**Pull request by kaspersorensen**
Do not merge yet - This PR is so far only here to remind us of ongoing work in the branch and to get updates from Travis.

This is a rework of the configuration classes of DataCleaner - see #363

### CODE REVIEW

**Comment by kaspersorensen** [comment]
Note to self: Make sure to solve the resource type handler... It should be able to load resources from the new home directory and then we can avoid a lot of hacks in monitor to make resource loading work per-tenant.

File: engine/core/src/main/java/org/datacleaner/configuration/DataCleanerConfigurationImpl.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    public static ReferenceDataCatalog defaultReferenceDataCatalog() {
+        return new ReferenceDataCatalogImpl();
+    }
+
+    public static DatastoreCatalog defaultDatastoreCatalog() {
+        return new DatastoreCatalogImpl();
+    }
+
+    public static RepositoryFolder defaultHomeFolder() {
+        return new FileRepository(".");
```
Comment (tomaszguzialek) [inline engine/core/src/main/java/org/datacleaner/configuration/DataCleanerConfigurationImpl.java]: So the default is still application directory? Did not decide to kill this behaviour?

Comment (kaspersorensen) [inline engine/core/src/main/java/org/datacleaner/configuration/DataCleanerConfigurationImpl.java]: It is still the default if you just instantiate a new DataCleanerConfigurationImpl. This is still fine behaviour for casual use of the class (say, in a unittest or so) but in the two "real" environments ("desktop" and "monitor") we set another home folder that fits those situations.

File: engine/env/cluster/src/test/java/org/datacleaner/cluster/DistributedAnalysisRunnerTest.java [hunk 2]
```diff
@@ -118,7 +118,7 @@ public void testErrorHandlingFourSlaves() throws Exception {
 
     public void testUndistributableAnalyzer() throws Exception {
         final Datastore datastore = TestHelper.createSampleDatabaseDatastore("orderdb");
-        final AnalyzerBeansConfiguration configuration = new AnalyzerBeansConfigurationImpl()
+        final DataCleanerConfiguration configuration = new AnalyzerBeansConfigurationImpl()
```
Comment (tomaszguzialek) [inline engine/env/cluster/src/test/java/org/datacleaner/cluster/DistributedAnalysisRunnerTest.java]: Why are we using the old implementation in these tests?

**Comment by tomaszguzialek** [comment]
Just asked two questions (one already answered), the rest looks good... Although the diff was unhuman ;).

cat_toilet_paper

File: engine/env/cluster/src/test/java/org/datacleaner/cluster/DistributedAnalysisRunnerTest.java (hunk 2 above)
Comment (kaspersorensen) [inline engine/env/cluster/src/test/java/org/datacleaner/cluster/DistributedAnalysisRunnerTest.java]: I guess it's one of many cases where the old stuff is still in use. I didn't get around to fix ALL the deprecation warnings.

**Pull request merged**